lib/active_model/serializer/adapter/json.rb

```diff
@@ -4,44 +4,122 @@ class ActiveModel::Serializer::Adapter::Json < ActiveModel::Serializer::Adapter
 
         def serializable_hash(options = nil)
           options ||= {}
+          hash = {}
+
           if serializer.respond_to?(:each)
-            result = serializer.map { |s| FlattenJson.new(s).serializable_hash(options) }
+            hash = serialize_array_without_root(serializer, options)
           else
-            hash = {}
+            hash = resource_object_for(serializer)
+            hash = add_resource_relationships(hash, serializer)
+          end
+
+          { root => hash }
+        end
+
+        def fragment_cache(cached_hash, non_cached_hash)
+          ActiveModel::Serializer::Adapter::Json::FragmentCache.new().fragment_cache(cached_hash, non_cached_hash)
+        end
+
+        private
 
-            core = cache_check(serializer) do
-              serializer.attributes(options)
+        # iterate through the associations on the serializer,
+        # adding them to the parent as needed (as singular or plural)
+        #
+        # nested_associations is a list of symbols that governs what
+        # associations on the passed in seralizer to include
+        def add_resource_relationships(parent, serializer, nested_associations = {})
+          # have the include array normalized
+          nested_associations = ActiveModel::Serializer::Utils.include_args_to_hash(nested_associations)
+
+          included_associations = if nested_associations.present?
+                                    serializer.associations.select { |association|
```

Comment:
I don't like the massive indentation. I think I'm going to go back to wrapping the whole thing in parenthesis, or see if I can make this a method.